### PR TITLE
Media: Update `get_post_galleries()` to return Gallery Blocks.

### PR DESCRIPTION
**Pull request by costdev**
Galleries in Gutenberg don't use the `[gallery]` shortcode, so they are not returned by `get_post_galleries()`.

This change makes `get_post_galleries()` also check for Gallery blocks and return them.

Trac ticket: https://core.trac.wordpress.org/ticket/43826

### Tasks

**src/wp-includes/media.php**
- [x] Add original implementation.
- [x] Add new implementation.
- [x] Review and iterate for performance.
- [x] Ensure that all tests pass.

**tests/phpunit/media.php**
- [x] Tests: Reduce the use of `assertEquals()` in favour of stricter options.
- [x] Tests: Correct the placement of `expected` and `actual` values in tests added in this PR.
- [x] Tests: Move all `get_post_galleries()` tests to a separate file: `tests/phpunit/tests/media/getPostGalleries.php`.

**tests/phpunit/media/getPostGalleries.php**
- [x] Tests: Add appropriate `@covers` annotation to test class.
- [x] Tests: Add appropriate `@group` annotations for each test method.
- [x] Tests: Correct the placement of `expected` and `actual` values in pre-existing test methods.
- [x] Tests: Add `message` parameters to test methods with multiple assertions.
- [x] Tests: Rename test methods for clarity when running the tests.
- [x] Tests: Improve docblock comments for each test method.
- [x] Tests: Change `assertSame()` to `assertSameSetsWithIndex()`.
- [x] Tests: Add assertions to prevent fatal errors.
- [x] Tests: Add more tests to improve coverage for `get_post_galleries()`.

~~Tests: Reduce the number of test methods by using data providers where possible.~~
Data providers deferred until a later review of the test suite.

### CODE REVIEW

File: tests/phpunit/tests/media.php [hunk 1]
```diff
@@ -707,6 +707,355 @@ public function test_post_gallery_images() {
...
+		$post_id = $this->factory->post->create(
+			array(
+				'post_content' => '<!-- wp:gallery -->',
+			)
+		);
+
+		$galleries = get_post_galleries( $post_id, false );
+
+		$this->assertTrue( is_array( $galleries ) );
+		$this->assertEmpty( $galleries[0]['src'] );
```
Comment (costdev) [inline tests/phpunit/tests/media.php:734]: The error in the tests occurs here.

And this is the culprit in source:
https://github.com/WordPress/wordpress-develop/blob/7dc550a619576f0a7e708086e11f2f1637a011fa/src/wp-includes/media.php#L4850

If this line is moved outside of `if ( ! empty( $src ) ) {}`, then `test_mixed_post_galleries()` and `test_mixed_post_galleries_data()` fail as the `Actual` arrays contain an extra `array( 'src' => array() )` element.

See the only other `assertEmpty()` check on this in the test suite [here](https://github.com/costdev/wordpress-develop/blob/f7d670c767254e58a52a9ae1d4b9921678686458/tests/phpunit/tests/media.php#L582), which tests an empty `[gallery]` shortcode.

This suggests that it should return `array( 'src' => array() )` if no `src` is found.

From the [Docs](https://developer.wordpress.org/reference/functions/get_post_galleries/) on this method:
> Return
> (array) A list of arrays, each containing gallery data and srcs parsed from the expanded shortcode.

So, _should_ it return `array( 'src' => array() )` if no `src` is found?
- If so, then  `test_mixed_post_galleries()` and `test_mixed_post_galleries_data()` need to be updated to account for this extra element.
- If not, then the `$this->assertEmpty( $galleries[0]['src'] );` assertion above should be removed.
- If neither, I'm not sure where the method goes wrong.

Comment (costdev) [reply tests/phpunit/tests/media.php:734]: `parse_blocks()` detects a shortcode as a block with a `null` value for `blockName`.
As shortcodes are handled by the code above these changes, the following guard is now used:
```
if( ! $block['blockName'] ) {
	continue;
}
```

**Comment by glendaviesnz** [comment]
This tested well for me on a site running the Arbutus theme which uses `get_post_galleries` to generate gallery excerpts. This updated call worked for a v1 and v2 gallery block, a gallery blocked nested 2 deep in  group blocks, and also for a short code gallery:
Screen Shot 2021-11-16 at 11 00 31 PM

LGTM once the failing tests are sorted.

File: src/wp-includes/media.php [hunk 2]
```diff
@@ -4767,6 +4767,95 @@ function get_post_galleries( $post, $html = true ) {
...
+					'src' => $srcs,
+				);
+
+				continue;
+			}
+
+			// Otherwise, extract srcs from the innerHTML.
+			preg_match_all( '#src=([\'"])(.+?)\1#is', $block['innerHTML'], $src, PREG_SET_ORDER );
+
+			if ( ! empty( $src ) ) {
```
Comment (hellofromtonya) [inline src/wp-includes/media.php:4847]: Why `empty()` here? `preg_match_all()` will return an `array()`, unless there's a failure and then it returns `false` on failure (though not sure when `false` will be thrown). The guard here is protecting the `foreach` which requires an iterable.

Changing the guard to `if ( is_array( $src ) )` is more clear why the guard exists and what it's doing. Then if `$src` is an empty array, the `foreach` doesn't iterate, skips it.

File: src/wp-includes/media.php [hunk 3]
```diff
@@ -4767,6 +4767,95 @@ function get_post_galleries( $post, $html = true ) {
...
+				foreach ( $src as $s ) {
+					if ( isset( $s[2] ) && ! in_array( $s[2], $srcs, true ) ) {
```
Comment (hellofromtonya) [inline src/wp-includes/media.php:4849]: If `$src` is renamed to something like `$found_srcs`, the `$s` can be renamed to `$src` to improve readability.

File: src/wp-includes/media.php [hunk 4]
```diff
@@ -4767,6 +4767,95 @@ function get_post_galleries( $post, $html = true ) {
...
+
+			// New Gallery block format as an array.
+			if ( $has_inner_blocks ) {
+				$attrs = wp_list_pluck( $block['innerBlocks'], 'attrs' );
+				$ids   = wp_list_pluck( $attrs, 'id' );
+
+				foreach ( $ids as $id ) {
+					$url = wp_get_attachment_url( $id );
+
+					if ( $url && ! in_array( $url, $srcs, true ) ) {
```
Comment (hellofromtonya) [inline src/wp-includes/media.php:4804]: `wp_get_attachment_url()` returns a string if URL found, else `false` (though filtered). As it's filtered, could be anything coming back including something that passes the truthy check.

@costdev what about changing this to `is_string( $src )` to ensure it's a string data type?

```suggestion
					if ( is_string( $url ) && ! in_array( $url, $srcs, true ) ) {
```

Here's the benchmark on truthy vs `is_string()` https://3v4l.org/IGHWg#v7.4.25. Truthy is slightly faster on average at 0.0028 μs. But the `is_string()` is more clear and guards the `in_array()` to do its thang only if a string URL is found. Thoughts?

File: src/wp-includes/media.php [hunk 5]
```diff
@@ -4767,6 +4767,95 @@ function get_post_galleries( $post, $html = true ) {
...
+				$galleries[] = array(
+					'ids' => implode( ',', $ids ),
+					'src' => $srcs,
+				);
+
+				continue;
+			}
+
+			// Old Gallery block format as HTML.
+			if ( $html ) {
```
Comment (hellofromtonya) [inline src/wp-includes/media.php:4818]: Same comment https://github.com/WordPress/wordpress-develop/pull/1864/files#r750309682

File: src/wp-includes/media.php [hunk 6]
```diff
@@ -4767,6 +4767,95 @@ function get_post_galleries( $post, $html = true ) {
...
+				$galleries[] = array(
+					'ids' => implode( ',', $ids ),
+					'src' => $srcs,
+				);
+
+				continue;
+			}
+
+			// Otherwise, extract srcs from the innerHTML.
+			preg_match_all( '#src=([\'"])(.+?)\1#is', $block['innerHTML'], $src, PREG_SET_ORDER );
```
[the remark this answers is not in the document]
Comment (costdev) [reply src/wp-includes/media.php:4845]: Agreed! I'll do that in the next iteration.

File: src/wp-includes/media.php (hunk 3 above)
Comment (costdev) [reply src/wp-includes/media.php:4849]: Agreed! I'll do that in the next iteration.

File: src/wp-includes/media.php [hunk 7]
```diff
@@ -4767,6 +4767,95 @@ function get_post_galleries( $post, $html = true ) {
...
+			}
+
+			// All blocks nested inside non-Gallery blocks should be in the root array.
+			if ( $has_inner_blocks && 'core/gallery' !== $block['blockName'] ) {
+				array_push( $post_blocks, ...$block['innerBlocks'] );
+				continue;
+			}
+
+			// New Gallery block format as HTML.
+			if ( $has_inner_blocks && $html ) {
```
[the remark this answers is not in the document]
Comment (costdev) [reply src/wp-includes/media.php:4788]: `$html` is `bool` type.

File: src/wp-includes/media.php (hunk 4 above)
Comment (costdev) [reply src/wp-includes/media.php:4804]: Sounds great, and the speed difference is minimal to bring a better check. Interestingly, `false !== $url` is slower than both of them. I'll add `is_string( $url )` to the next iteration.

File: src/wp-includes/media.php (hunk 5 above)
Comment (costdev) [reply src/wp-includes/media.php:4818]: `$html` is `bool` type.

File: src/wp-includes/media.php [hunk 8]
```diff
@@ -4767,6 +4767,95 @@ function get_post_galleries( $post, $html = true ) {
...
+			if ( $html ) {
+				$galleries[] = $block['innerHTML'];
+				continue;
+			}
+
+			// Old Gallery block format as an array.
+			$ids = ! empty( $block['attrs']['ids'] ) ? $block['attrs']['ids'] : array();
+
+			// If present, use the image IDs from the JSON blob as canonical.
+			if ( ! empty( $ids ) ) {
```
[the remark this answers is not in the document]
Comment (costdev) [reply src/wp-includes/media.php:4827]: This guard is needed to differentiate between an old gallery that has a JSON blob and one that doesn't.

- An old gallery that has a JSON blob not only has the URLs retrieved by `wp_attachment_url()`, but also has the `ids` added to its element in the `$galleries` array. The guard prevents `implode()` being called on an empty array and prevents the `src` key of the gallery being set to the default value of `$srcs`; an empty array.

- An old gallery that doesn't have a JSON blob must have the URLs pulled from the `src` attributes within the HTML.

File: src/wp-includes/media.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (hellofromtonya) [reply src/wp-includes/media.php:4788]: Aha. Then disregard my comment. Thanks Colin.

File: src/wp-includes/media.php (hunk 5 above)
Comment (hellofromtonya) [reply src/wp-includes/media.php:4818]: `$html` is confusing. `$is_html` is more descriptive but that's a refactor of the original code. No need to change in this ticket.

File: src/wp-includes/media.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (hellofromtonya) [reply src/wp-includes/media.php:4827]: Aha, yes, the `implode()` is outside of the `foreach()`. You're right.

File: src/wp-includes/media.php (hunk 2 above)
Comment (hellofromtonya) [reply src/wp-includes/media.php:4847]: >As preg_match_all() returns array( array() ) when no matches are found, this guard should actually be checking $src[0]. I'll update the guard in the next iteration.

I agree. Good job!

**Review by hellofromtonya — APPROVED** [review]
Great job @costdev and @glendaviesnz 🎉 The PR to handle both v1 and v2 gallery block looks ready for `commit` 👍

**Comment by hellofromtonya** [comment]
Committed via changeset https://core.trac.wordpress.org/changeset/52190.

**Pull request closed without merging**